text/0000-tool-attributes.md

```diff
@@ -0,0 +1,262 @@
+
+### Attributes
+
+This section assumes that attributes (e.g., `#[test]`) have already been taught.
+
+You can use attibutes in your crate to pass information to tools. For now, this
+facility is limited to the tools we include with the Rust distribution.
+
+The names of these attributes are a path starting with the name of a tool, and
+then one or more identifiers, e.g., `#[tool_name::foo]` or
+`#[tool_name::bar::baz::qux(argument)]`. Such paths hide any attribute-like
+macros with the same name and location.
+
+For example, using `#[rustfmt::skip]` indicates that an item (such as a function)
+should not be formatted by Rustfmt:
+
+```
+#[rustfmt::skip]
+fn foo() { this_will_be_kept_as_is_by_rustfmt(); }
+
+fn bar() { this_will_be_reformatted }
+
+mod baz {
+    #![rustfmt::skip]
+    // Rustfmt will skip this whole module.
+}
+```
+
+### Lints
+
+This section assumes lints have already been taught.
+
+Lints can be defined hierarchically as a path, as well as just a single name.
+For example, `nonstandard_style::non_snake_case_functions` and
+`nonstandard_style::uppercase_variables`. Note this RFC is not proposing
+changing any existing lints, just extending the current lint naming system. Lint
+names cannot be imported using `use`.
+
+Lints can be enforced by tools other than the compiler. For example, Clippy
+provides a large suite of lints to catch common mistakes and improve your Rust
+code. Lints for tools are prefixed with the tool name, e.g., `clippy::box_vec`.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Long-term solution
+
+There will be some opt-in mechanism for crates to declare that they want to
+allow use of a tool's attributes. This might be in the source text (an attribute
+as in #1755 or new syntax, e.g., `extern attribute foo;`) or passed to rustc as
+a command line flag (e.g., `--extern-attr foo`). The exact mechanism is
+deliberately unspecifed.
+
+After opting-in to `foo`, a crate can use `foo` as the base of a path in any
+attribute in the crate. E.g., allowing `#[foo::bar]` to be used (but not
+`#[foo]`). This mechanism is follows the normal macro hygiene rules. Depending
+on the opt-in mechanism a tool might be able to specify to the compiler which
+paths are valid, e.g., allow `#[foo::bar]` but disallow `#[foo::baz]`. I would
+hope that we'd be able to reuse most of the macro naming feature (see #1561)
+here (i.e., this won't be a whole new specification, we'll just allow a new way
+to base paths).
+
+Unscoped attributes will be reserved for the language and can't be used by tools.
+
+During macro expansion, when faced with an attribute, the compiler first checks
+if the attribute matches any of the declared or built-in attributes. If not, it
+tries to find a macro using the [macro name resolution rules](https://github.com/rust-lang/rfcs/blob/master/text/1561-macro-naming.md).
+If this fails, then it reports a macro not found error. The compiler *may*
```

Comment:
This looks like an inconsistent prioritization to me.

It would be good if resolution for attribute paths worked consistently with the rest of resolution,
as if attribute names were just names defined in macro namespace with usual scoping, shadowing, etc.

In particular, I see parallels with built-in types, std prelude types and user defined types.
User defined macro attributes -> user defined types, found first by name resolution.
Standard attributes potentially expressible as standard library macros (e.g. potentially `#[test]`) -> std prelude (e.g. `Box`), found as a fallback if no user-defined names are found.
Built-in attributes (e.g. `#[repr]`) -> built-in types (e.g. `u8`), language prelude, found as a fallback if no std prelude names are found.

This way we can always introduce new built-in attributes without breaking user-defined macros, like it was done with `u128`/`i128` for types.

Now attributes for tools need to somehow be fitted into this scheme.
This test implies that they should shadow everything when enabled (e.g. through `--extern-attr`) and break macros. I'd actually expect the opposite priority and placing them into the language prelude for example.
(There shouldn't be conflicts there due to different number of path segments - 1 for built-in, >1 for extern).

Reply:
This seems like a good change, thanks!